testrepository/commands/run.py

```diff
@@ -169,29 +189,45 @@ def gather_errors(test_dict):
 
     def run(self):
         repo = self.repository_factory.open(self.ui.here)
-        if self.ui.options.failing or self.ui.options.analyze_isolation:
-            ids = self._find_failing(repo)
-        else:
-            ids = None
-        if self.ui.options.load_list:
-            list_ids = set()
-            # Should perhaps be text.. currently does its own decode.
-            with open(self.ui.options.load_list, 'rb') as list_file:
-                list_ids = set(parse_list(list_file.read()))
-            if ids is None:
-                # Use the supplied list verbatim
-                ids = list_ids
-            else:
-                # We have some already limited set of ids, just reduce to ids
-                # that are both failing and listed.
-                ids = list_ids.intersection(ids)
         if self.ui.arguments['testfilters']:
             filters = self.ui.arguments['testfilters']
         else:
             filters = None
-        testcommand = self.command_factory(self.ui, repo)
+        profiles = self.ui.options.profiles
+        if not isinstance(profiles, six.text_type):
+            profiles = profiles.decode('utf8')
+        profiles = profiles.split(_u(','))
+        if profiles == ['']:
+            profiles = None
+        testcommand = self.command_factory(self.ui, repo, profiles=profiles)
         testcommand.setUp()
         try:
+            if self.ui.options.failing or self.ui.options.analyze_isolation:
+                ids = self._find_failing(repo, testcommand.profiles)
+            else:
+                ids = None
+            if self.ui.options.load_list:
+                list_ids = set()
+                # Should perhaps be text.. currently does its own decode.
+                with open(self.ui.options.load_list, 'rb') as list_file:
+                    list_ids = set(parse_list(list_file.read()))
+                list_ids = apply_profiles(
+                    testcommand.default_profiles, list_ids)
+                if ids is None:
+                    # Use the supplied list verbatim
+                    ids = list_ids
+                else:
+                    # We have some already limited set of ids, just reduce to ids
+                    # that are both failing and listed.
+                    _ids = {}
+                    for test, id_meta in ids.items():
+                        if test not in list_ids:
+                            continue
+                        test_profiles = set(id_meta['profiles'])
+                        test_profiles.intersection_update(list_ids[test]['profiles'])
+                        _ids[test] = {'profiles': sorted(test_profiles)}
+                    ids = _ids
```

Comment:
This chunk should almost certainly be in its own method that returns what's now `ids`. This current method is ~140 lines long and is already stretching my (admittedly tired) brain.